### PR TITLE
Add support for transcoding plex streams before playback

### PR DESCRIPTION
**Pull request by chicco-carone**
# What does this implement/fix?
Add support into the plex provider to let the server transcode audio before sending it to MA to reduce streaming bandwidth limits. Plex supports calling its api from a different endpoint and specifying the bitrate and codec to stream from the server to avoid always streaming the original file, this pr adds into the provider settings an option to choose either the original quality or opus at a specific bitrate. To avoid issues when updating it defaults to original which was the previous behavior and then the user if needed can change it.

**Related issue (if applicable):**

This issue was discussed in the MA discord server in the following thread [Thread Link](https://discord.com/channels/753947050995089438/1522653891689316394)

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue) — `bugfix`
- [ ] New feature (non-breaking change which adds functionality) — `new-feature`
- [x] Enhancement to an existing feature — `enhancement`
- [ ] New music/player/metadata/plugin provider — `new-provider`
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected) — `breaking-change`
- [ ] Refactor (no behaviour change) — `refactor`
- [ ] Documentation only — `documentation`
- [ ] Maintenance / chore — `maintenance`
- [ ] CI / workflow change — `ci`
- [ ] Dependencies bump — `dependencies`

## Checklist

- [x] The code change is tested and works locally.
- [x] `pre-commit run --all-files` passes.
- [x] `pytest` passes, and tests have been added/updated under `tests/` where applicable.
- [ ] For changes to shared models, the companion PR in `music-assistant/models` is linked.
- [ ] For changes affecting the UI, the companion PR in `music-assistant/frontend` is linked.
- [x] I have read and complied with the project's [AI Policy](https://github.com/music-assistant/.github/blob/main/AI_POLICY.md) for any AI-assisted contributions.
- [ ] I have raised a PR against the documentation repository targeting the main or beta branch as appropriate.

### CODE REVIEW

**Comment by OzGav** [comment]
@anatosun

**Comment by anatosun** [comment]
Are you aware of [this thread](https://github.com/music-assistant/server/pull/3561)?

**Comment by chicco-carone** [comment]
I updated the pr so that if plex has the metadata to support transcoding the file it will transcode it else it will fallback the the previous behaviour for files that are not yet fully processed

**Comment by anatosun** [comment]
Good, have you tested the behavior on a newly added track that was not analyzed?

**Comment by chicco-carone** [comment]
I haven't been able since if I add new tracks plex process them instantly and I don't even have the time to go I mass to test it. I'll see if I can slow it down somehow on my server to test

**Comment by chicco-carone** [comment]
I'm completely out of ideas on how I could test it, I gave my plex container half cpu and imported a whole series with some music to hopefully slow it down but it doesn't even appear in mass in time. If you have any ideas on how I could test this I would be happy to try it to then close this but currently i'm not sure

**Comment by anatosun** [comment]
you should defer the analysis to a scheduled task...

image

**Comment by chicco-carone** [comment]
That analysis setting is only for the features used in plexamp like loudness normalization and djs and it doesn't change anything for mass since it requests the whole track the metadata that plex needs for transcoding is basically only missing when the track has no metadata (so like no album cover etc) so even with that turned off it still analyzes the metadata needed for transcoding and as such it isn't a reliable way to test it. with those 2 things disabled though it plays back fine even without analysis.
image

**Comment by anatosun** [comment]
This task also populates the size of the track, which the streaming process requires. Without it, you'll see that it raises an error and this is where it should fallback on downloading the track.

**Comment by chicco-carone** [comment]
This track i played it after disabling the analysis and adding it and it played back fine

**Comment by anatosun** [comment]
... and it works because you have `download=1` in the arguments

**Comment by chicco-carone** [comment]
In the provided example mass is playing back the track in opus which is correct per the settings that I set in the provider settings to transcode and its not falling back to `download=1` since in that case it should have gotten from the server the original flac file and the playback should have been in flac not opus. 
Regarding the analysis in plex like its mentioned here in the support articles https://support.plex.tv/articles/200289336-analyze-media/
 > Analysis is automatically performed when content is added to your Library.

And its separate from the sonic analysis that you can disable in the settings For what this pr covers the transcode works fine and if there is no metadata it fallbacks to the original file works fine

**Comment by anatosun** [comment]
My server does not behave this way. The size of the file is populated only after the scheduled analysis. 

But you seem to say that the fallback to the original could be tested?

**Comment by chicco-carone** [comment]
I'm not sure how to test this in my current environment since I can't reproduce the edge case that would cause this problem. That said here in the code where it decides to transcode or not if any of the checks in the if fail for any possible reason (like plex not having the metadata) it will fallback to using download=1 which is already proven by #3561 to work in those specific cases.
```python
if (quality_bitrate := self._get_stream_quality_bitrate()) and audio_stream:
    stream_details.path = self._get_transcode_url(plex_track, quality_bitrate)
    stream_details.stream_type = StreamType.HLS
    stream_details.audio_format.content_type = ContentType.OPUS
    stream_details.audio_format.bit_rate = quality_bitrate
    return stream_details

download_url = self._plex_server.url(f"{media_part.key}?download=1", True)
```

**Comment by chicco-carone** [comment]
I want to finish this but I can't understand what's the problem to get this merged

**Comment by OzGav** [comment]
Maybe you can mock a Plex track missing the analysis-derived metadata (no audio_stream / no size) and assert stream details fall back to the `download=1` URL with the original content type?

Also there are conflicts to be resolved and copilot comments.

File: music_assistant/providers/plex/__init__.py [hunk 1]
```diff
@@ -1083,10 +1113,60 @@ async def _cleanup_stale_library_mappings(self) -> None:
...
+            self._plex_server.url(
+                f"/music/:/transcode/universal/start.m3u8?{urlencode(params)}",
+                True,
+            )
+        )
+
     async def _run_async(
         self, call: Callable[Param, RetType], *args: Param.args, **kwargs: Param.kwargs
     ) -> RetType:
-        await self.get_myplex_account_and_refresh_token(str(self.get_setup_value(CONF_AUTH_TOKEN)))
```
Comment (MarvinSchenkel) [inline music_assistant/providers/plex/__init__.py:1103]: Is this call meant to be removed?

File: music_assistant/providers/plex/strings.json [hunk 2]
```diff
@@ -54,6 +54,17 @@
       "label": "Items per hub",
       "description": "Maximum number of items to load from each hub (default: 10)"
     },
+    "stream_quality": {
+      "label": "Stream quality",
+      "description": "Audio quality to request from Plex for music tracks. Original streams the source file directly. Lower bitrates ask Plex Media Server to transcode before sending audio to Music Assistant.",
```
[the remark this answers is not in the document]
Comment (MarvinSchenkel) [reply music_assistant/providers/plex/strings.json]: This is still valid feedback

**Comment by OzGav** [comment]
There are some comments from Marvin waiting. Just marking this as draft so we know what to work on. Please mark as ready for review when you want another look.

**Comment by chicco-carone** [comment]
I addressed all the remaining issues and the merge conflict so now the pr should be ready for final review

**Review by OzGav — APPROVED** [review]
LGTM but will just let Marvin have a final look

**Comment by MarvinSchenkel** [comment]
LGTM, let's test it in beta. Thanks @chicco-carone
